### PR TITLE
feat: subtract `IntrinsicGas` before executing a transaction

### PR DESCRIPTION
**Pull request by Flouse**
**This PR:**
- check the gas limit and return an error if the gas limit is less than IntrinsicGas
- [feat: subtract IntrinsicGas before executing a tx](https://github.com/nervosnetwork/godwoken-polyjuice/pull/160/commits/2a644bfd6c1d9a14c5ba3216578ccac99e0611e7)
- make sure gas_used not exceed msg.gas

## References
- https://github.com/nervosnetwork/godwoken-web3/issues/335
- https://github.com/ethereum/go-ethereum/blob/b3b8b268eb585dfd3c1c9e9bbebc55968f3bec4b/params/protocol_params.go#L32-L34

### CODE REVIEW

File: c/polyjuice.h [hunk 1]
```diff
@@ -1366,27 +1373,36 @@ int run_polyjuice() {
...
     debug_print_int("evmc_result.output_size", res.output_size);
     // The output contains data coming from REVERT opcode
     debug_print_data("evmc_result.output_data:", res.output_data,
                      res.output_size > 100 ? 100 : res.output_size);
   }
-  
+
+  debug_print_int("[run_polyjuice] gas left", res.gas_left);
+  uint64_t gas_used =
+      (uint64_t)(res.gas_left <= 0 ? initial_gas : initial_gas - res.gas_left);
```
Comment (TheWaWaR) [inline c/polyjuice.h:1395]: Shouldn't we report error when `res.gas_left < 0`?

Comment (Flouse) [reply c/polyjuice.h:1395]: https://github.com/nervosnetwork/godwoken-polyjuice/blob/159378db76979bfd92f11cfc33bbd6e45d01941d/c/polyjuice.h#L1024-L1028

The transaction will turn out to be a failed transaction with EVMC_OUT_OF_GAS exit_code.
And then, Godwoken will [handle tx fee here](https://github.com/nervosnetwork/godwoken/blob/76b43a1762f5cb703aeeafc4a31ec67e5d543bb5/crates/generator/src/generator.rs#L677-L711).

Note: In the latest version of Godwoken, failed transactions are packed into Layer2 block.

**Review by TheWaWaR — APPROVED** [review]

**Review by magicalne — APPROVED** [review]

**Review by RetricSu — APPROVED** [review]

File: c/polyjuice.h [hunk 2]
```diff
@@ -1410,16 +1426,10 @@ int run_polyjuice() {
     return clean_evmc_result_and_return(&res, -1);
   }
   if (msg.gas < res.gas_left) {
```
Comment (RetricSu) [inline c/polyjuice.h:1428]: question: when will such a condition happen? (what does this condition mean)

**Review by jjyr — DISMISSED** [review]

File: c/polyjuice.h (hunk 2 above)
Comment (Flouse) [reply c/polyjuice.h:1428]: It was some debugging code.
will remove this unreachable if branch.

File: c/polyjuice_utils.h [hunk 3]
```diff
@@ -318,4 +318,42 @@ bool is_evmc_error(int error_code) {
...
+  if (msg->input_size > 0) {
+    // Zero and non-zero bytes are priced differently
+    uint64_t non_zero_bytes = 0;
+    for (size_t i = 0; i < msg->input_size; i++) {
+      if (*(msg->input_data + i) != 0)
+        non_zero_bytes++;
+    }
+
+    // Make sure we don't exceed uint64 for all data combinations
+    if ((UINT64_MAX - *min_gas) / DATA_NONE_ZERO_TX_GAS < non_zero_bytes) {
```
Comment (jjyr) [inline c/polyjuice_utils.h:344]: @TheWaWaR Please help to check this line. I don't quietly understand how it works.

Comment (magicalne) [reply c/polyjuice_utils.h:344]: It calculates how many non-zero bytes can use by given such gas. It looks good.
I think we can use constants here to replace (UINT64_MAX - *min_gas) / DATA_NONE_ZERO_TX_GAS.

Comment (Flouse) [reply c/polyjuice_utils.h:344]: if ((UINT64_MAX - *min_gas) / DATA_NONE_ZERO_TX_GAS < non_zero_bytes)
then *min_gas += non_zero_bytes * DATA_NONE_ZERO_TX_GAS; 

*min_gas -> overflow(UINT64_MAX)

**Review by magicalne — APPROVED** [review]

**Review by TheWaWaR — APPROVED** [review]

**Pull request merged**